PowerEditor/src/MISC/Common/FileInterface.cpp

```diff
@@ -36,7 +28,9 @@ Win32_IO_File::Win32_IO_File(const wchar_t *fname)
 		std::wstring fn = fname;
 		std::wstring_convert<std::codecvt_utf8_utf16<wchar_t>> converter;
 		_path = converter.to_bytes(fn);
-		_hFile = ::CreateFileW(fname, _accessParam, _shareParam, NULL, _dispParam, _attribParam, NULL);
+
+		DWORD dispParam = ::PathFileExistsW(fname) ? TRUNCATE_EXISTING : CREATE_ALWAYS;
+		_hFile = ::CreateFileW(fname, _accessParam, _shareParam, NULL, dispParam, _attribParam, NULL);

```

Comment:
I do not like that this is a race condition now. What if the file does not exist when `PathFileExistsW` is called but exists when `CreateFileW` is called? You then are opening the file with the `CREATE_ALWAYS` option and possibly kill the ADS.
You could change your code to first try to open the file with `TRUNCATE_EXISTING` and if that fails with last error = `ERROR_FILE_NOT_FOUND ` you then go on and try open with `CREATE_NEW`.
If now that fails with last error = `ERROR_FILE_EXISTS` you could try  `TRUNCATE_EXISTING` again.
Do those two methods in a loop with like 5 tries and if all fail then report save failure to user.
This guarantees we never loose ADS and we always have a truncated file after open.

Reply:
Good point. But I think one set of TRUNCATE_EXISTING - CREATE_NEW - TRUNCATE_EXISTING would suffice.
Thanks.

Reply:
@pnedev 
I see. Thank you for the explanation.
With the same logic, such race condition could happen anytime: a program can delete file just after CreateFileW is called. 
The race condition I do care about is in the same process, other program which can alter the same files opened in Notepad++ is out of our control. 
The problem what @Klotzi111 has pointed out could happen, but it happens very rarely. If we want to prevent it, we should enhance the code in Notepad++ everywhere and it would be not even efficient for all we had done.

Reply:
@donho ,

> With the same logic, such race condition could happen anytime. For exemple: a program can delete file just after CreateFileW is called.

That is not completely the case.
After CreateFileW() we already have the valid file handle and we also do not allow file deletion access (the sharing options used with CreateFileW do not include FILE_SHARE_DELETE).

> The race condition I do care about is in the same process, other program which can alter the same files opened in Notepad++ is out of our control.

This is also not quite true. Considering such scenarios allows us to take appropriate measures and we are responsible for the proper handling even if other processes/programs might be involved.
For example the OS provides the LockFile interface which idea is to properly synchronize OS resources shared by different programs.

> If we want to prevent it, we should enhance the code in Notepad++ everywhere and it would not be even efficient for all we had done.

Well, it is better to make the new code portions behave properly (to best of our knowledge and considerations) even-though there might be portions in the existing code that are prone to similar problems.
If we have a hole in our boat's bottom let's not puncture another one.

There is no functional difference between your current changes and using CreateFileW with TRUNCATE_EXISTING directly.

Thanks.

Reply:
@pnedev 

> After CreateFileW() we already have the valid file handle and we also do not allow file deletion access (the sharing options used with CreateFileW do not include FILE_SHARE_DELETE).

Makes sense. I wasn't aware of it.

> If we have a hole in our boat's bottom let's not puncture another one.

Love this quote and totally agree with it.

>  Good point. But I think one set of TRUNCATE_EXISTING - CREATE_NEW - TRUNCATE_EXISTING would suffice.
Thanks.

I don't really get your idea. Could you create a new PR from the master please? I then will close this one.

Reply:
> @Klotzi111 ,
Please confirm you are currently working on a fix, thank you.

Confirmed. But please give me some days. (Much work at work 😄 )

Reply:
@pnedev 
According to our discussion here:
https://github.com/notepad-plus-plus/notepad-plus-plus/pull/14029#discussion_r1305584905

Please reconsider my request **Could you create a new PR from the master please? I then will close this one.**, or provide me the code that you feel suitable for this PR, in the comment.

Thank you.

Reply:
@donho , All
I have made a quick PR as discussed - haven't had time to verify it, sorry:
https://github.com/notepad-plus-plus/notepad-plus-plus/pull/14051

Please feel free to modify it as you like.